### DISCUSSION TITLE
[Q&A] How to run simulation in different device (10 raspberry pi for example)

### DISCUSSION THREAD

**Discussion by plaipmc** [NVIDIA/NVFlare#2513] [Q&A]
### Python version (`python3 -V`)

3.10.12

### NVFlare version (`python3 -m pip list | grep "nvflare"`)

2.4.0

### NVFlare branch (if running examples, please use the branch that corresponds to the NVFlare version, `git branch`)

2.4.0rc2

### Operating system

Ubuntu 22.04.3

### Have you successfully run any of the following examples?

- [X] hello-numpy-sag with simulator
- [X] hello-pt with simulator
- [X] hello-numpy-sag with POC
- [X] hello-pt with POC

### Please describe your question

I done run simulation (using simulator) with 4 clients and each of them have distinct dataset. I want to try on raspberry PI cluster and I'm not really know a lot on how to transfer from simulator within 1 device to many devices. Please recommend me. Thank you

**Comment by YuanTingHsieh** [comment]
Hi @plaipmc in your case you can try with our provision mode.
You modify the project.yml to specify 4 clients and then copy the startup kit folder to each client.
run "./startup/start.sh" for each client.
run "./startup/start.sh" for the server.
then you should be able to use the admin cli to communicate with the whole system.

Please find more information here: 
https://nvflare.readthedocs.io/en/stable/real_world_fl.html

**Reply by plaipmc** [reply]
I try it with just the example :

1. Create provision.yaml using `nvflare provision` select option 2 and didn't edit anything
2. Try on my machine which it can't (same as #2527)
3. Move on to many clients by copy startup folder to one device I access using ssh
4. I try to run "start.sh" but it said that I need to use only Python version 3.8 3.9 and 3.10. 3.11 didn't support yet.

I try to change the Python version but ended up with unupdateable device. I try to use virtual environment but it said "Unable to locate package python3.10-venv" and other version too. Do you have some recommendations?

**Reply by plaipmc** [reply]
Here is the result after using . start.sh

> WORKSPACE set to /home/tem/NVFlare/examples/provision/startup/..
PYTHONPATH is /local/custom:
start fl because of no pid.fl
new pid 122302
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/tem/NVFlare/examples/nvflare_example/lib/python3.11/site-packages/nvflare/private/fed/app/client/client_train.py", line 184, in <module>
    version_check()
  File "/home/tem/NVFlare/examples/nvflare_example/lib/python3.11/site-packages/nvflare/private/fed/app/utils.py", line 98, in version_check
    raise RuntimeError("Python versions 3.11 and above are not yet supported. Please use Python 3.8, 3.9 or 3.10.")
RuntimeError: Python versions 3.11 and above are not yet supported. Please use Python 3.8, 3.9 or 3.10.
Gracefully shutdown.
[1]+  Done                    $DIR/sub_start.sh

**Reply by chesterxgchen** [reply]
to simulate locally with different clients, try to use POC command

```nvflare poc prepare -n 3```

if will generate the project.yml you need. you can review the resulting project.yml

if you are ready for real-setup/deployment, not just local host, you can use the 

nvflare provision -p project.yml 

command to provision different startup kits and put the startup kit into each device

**Reply by YuanTingHsieh** [reply]
@plaipmc I don't have a rasp pi board.
What I would suggest:
1. Find out how to install Python 3.10 in rasp pi
2. Make sure each of your rasp pi board can see the server address (for example "hello.com")
    One way is to edit the /etc/hosts in each rasp pi client
3. Install nvflare on all clients and server
4. Run hello-numpy example with simulator on all clients and server
5. Run nvflare provision on the server site and copy the client startup kits to each rasp pi client
6. start server 
7. start client
8. start admin CLI to confirm that the system is up and running (check_status server and check_status client)
9. submit a job to the system by (submit_job xxx)
